### ISSUE TITLE
Missing display of changed values in another color (red) after stepping

### ISSUE THREAD

**Issue by MBilderbeek** [openMSX/openMSX#1657]
This was a popular feature of the Qt based debugger. A little tricky to implement in ImGUI, as it requires a specific caching for all the values that we want to show with this feature, in order to know whether it changed.

**Comment by MBilderbeek** [comment]
@m9710797 perhaps this could be implemented by using something similar to 'tracked RAM'. Call it a 'tracked debuggable'. I.e. the debuggable state gets remembered at each break state, so that at break N you can compare what changed compared to break N-1. That means, not on a GUI level but in the openMSX core. The GUI can then show the difference between debuggables. All it takes is extra memory to remember the previous debuggable state, but the GUI code shouldn't get complicated and does no caching. And the cache is only updated after leaving any breaked state.

Is that feasible?

**Comment by m9710797** [comment]
It's possible that in the past I said something different, but today I think this would be best solved at the GUI level, not in the openMSX core.
* E.g. many of our "debuggables" internally don't even have an array-like structure (e.g. the CPU registers, or the "romblocks" debuggable). So forcing those to use a common "TrackedDebuggable" backing store is not a good idea IMHO.
* I also still believe it's important to implement our debugging infrastructure with no or as little overhead as possible (when that debug feature is not in use).
* E.g. our "TrackedRAM" class has (a bit) of overhead on every write access. But we only use it sporadically (e.g. not for main RAM, and certainly not for CPU registers).

But maybe a copy of the full debuggable (e.g. the CPU registers and the open hex editors), each time we leave the breaked state, should be acceptable.

Some practical questions though:
* On drawing the debugger GUI we can compare the current with the snapshoted value and change the drawing color (to red) if different. But should we do that always, or only when again in breaked state?
* E.g. when continue to run (not single stepping) the registers and RAM content will more and more start to diverge from the snapshoted values. Should we then gradually color more and more in red? This might look cool, but in the end everything will look red.
* And occasionally some values may return back to the original. Making the colors flip between red and white.

Some wild ideas (not for the first implementation, or ever):
* Making a snapshot when leaving the breaked state is a good idea, and should be the default. But other possibilities are:
  * Allow to manually make a snapshot.
  * Make snapshot after a configurable time interval (realtime or emutime).
  * ...
* We could constantly color differences red, or only when breaked.

**Comment by MBilderbeek** [comment]
Good questions on the behavior! I'll try to get some input from frequent users of the debugger. E.g. @sdsnatcher @grauw @gilbertfrancois - what do you think? 

If things get really hard, we can always fall back to the simplest scenario: mimick what was in the old debugger. So no real-time updates of 'red' stuff, only show red for changed values when being in 'breaked' state.

**Comment by thegeps** [comment]
I'd prefere real-time red stuff update

**Comment by m9710797** [comment]
I agree that real-time updating the changes may look nice, and can even be useful in some cases.
I'm only afraid that when you're doing other tasks than actively debugging (so not frequently breaking and stepping through the code) then after a while almost everything will look red.
So I think I'll make it an option (and default disabled?).

**Comment by gilbertfrancois** [comment]
To me, highlighting changed values would make sense after every step in the program counter, and in the next count, they would be coloured back to default colour, unless they change again.  I am most interested in seeing the values that changed due to the last instruction. This would also avoid ending up with a "all red memory view", as @m9710797 mentioned above. I don't know how memory state of the MSX is stored in openMSX. In case of memory, I could imagine doing a diff on the previous state and current state of the memory (and registers?). All values that have changed in that step could be highlighted. But since I don't know if there is already something like this in the core or UI, I cannot estimate how much effort it would take to implement it.

On the other hand, for me the feature would be a 'nice to have', rather than a 'must have'.

**Comment by gilbertfrancois** [comment]
I forgot to mention that most of the time I look at changed values in break / stepping state, not during runtime at full speed.

**Comment by thegeps** [comment]
An option is the best choice. So IF needed it can be enabled. Anyway, when debugging, i generally check portions of RAM (for example) checking changes applied by a specific part of code, so maybe a table for wich I expect some specific behaviour so also runtime turning red is good to see

**Comment by m9710797** [comment]
I implemented this. Please test and give feedback.
